### PR TITLE
fix(Dropdown): fix `minCharacter` prop behaviour

### LINKED ISSUE #1718: Dropdown: minCharacters breaks search input
I believe this issue is caused by #1690
When minCharacters is set to 1 (by default) - it is not possible to delete the last character in search query.
When minCharacters is set to > 1 - it is not possible to enter the search query at all.

### Steps
* Create a dropdown with default minCharacters
* Create a dropdown with minCharacters = 2
* Click on dropdown and start typing search query (e.g. -> "Afgan")
* Now you want to change your search query to "USA", so delete first query and start typing "USA"
### Expected Result
* It should be possible to type in and change search query
### Actual Result
* when minCharacters is set to 1 (by default) it is not possible to delete the last character, it always remains in the input
* when minCharacters is set to > 1 it is not possible to enter the search query at all
### Version
0.68.4

### Testcase
https://codepen.io/anon/pen/PmMPMN

@dyesseyumba would you like to check this?

### PR DESCRIPTION
**Pull request by layershifter**
Fixes #1718.

# Behaviours

This PR fixes issues with `minCharacters` and also some other inconsistent things.

#### TODO

- [x] onClick
- [x] onClick (toogle)
- [x] onFocus
- [x] onChange
- [x] onChange (`Enter`)

> onFocus (placeholder #1299, will be done in separate task)

## onClick

_Note_, that if you click on input after it was openned, you will not close it, e.g. click don't toggle searchable dropdown.

#### `minCharacters` = 1

peek 2017-06-01 16-41

#### `minCharacters` = 3

Dropdown wasn't openned after click.

peek 2017-06-01 16-49

#### `minCharacters` = 1

_Note_, focus doesn't affect placeholder. After focus was lost first item was selected, because item was focused.

#### `minCharacters` = 1

peek 2017-06-01 16-47

#### `minCharacters` = 3

Dropdown wasn't openned after click.

peek 2017-06-01 16-50

## onChange

Note, is you have _no results message_ and you click `Enter` key, nothing will be happened.

#### `minCharacters` = 1

peek 2017-06-01 16-52

#### `minCharacters` = 3

peek 2017-06-01 16-53

### CODE REVIEW

File: docs/app/Examples/elements/Container/Types/ContainerExampleContainer.js [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+   { key: 'ae', value: 'ae', flag: 'ae', text: 'United Arab Emirates' },
+   { key: 'us', value: 'us', flag: 'us', text: 'United States' },
+ ]
 
-const ContainerExampleContainer = () => (
-  <Container>
-    <p>Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa strong. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede link mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi.</p>
-  </Container>
+const DropdownExampleSearchSelection = () => (
```
Comment (levithomason) [inline docs/app/Examples/elements/Container/Types/ContainerExampleContainer.js]: Whoops, this looks like the wrong content for this file.

Comment (levithomason) [reply docs/app/Examples/elements/Container/Types/ContainerExampleContainer.js]: Whoops, looks like I missed the big fat `WIP` 😊

Comment (layershifter) [reply docs/app/Examples/elements/Container/Types/ContainerExampleContainer.js]: Yep, I will revert this before merge :smile:

**Comment by layershifter** [comment]
@levithomason  I done everything except placeholder's behaviour. Let me know what you think about proposed changes in this PR, gifs were captured from SUI, you can compare them with our current docs. If it will okay, I'll add tests and prepare PR for merge.

File: src/modules/Dropdown/Dropdown.js [hunk 2]
```diff
@@ -332,19 +332,19 @@ export default class Dropdown extends Component {
   static defaultProps = {
     additionLabel: 'Add ',
     additionPosition: 'top',
+    closeOnBlur: true,
     icon: 'dropdown',
+    minCharacters: 1,
     noResultsMessage: 'No results found.',
+    openOnFocus: true,
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:341]: Only prop sort there.

File: src/modules/Dropdown/Dropdown.js [hunk 3]
```diff
@@ -418,9 +418,11 @@ export default class Dropdown extends Component {
     if (!prevState.focus && this.state.focus) {
       debug('dropdown focused')
       if (!this.isMouseDown) {
-        const { openOnFocus } = this.props
+        const { minCharacters, openOnFocus, search } = this.props
+        const openable = !search || (search && minCharacters === 1)
+
         debug('mouse is not down, opening')
-        if (openOnFocus) this.open()
+        if (openOnFocus && openable) this.open()
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:427]: Defines behaviour with `focus`.

File: src/modules/Dropdown/Dropdown.js [hunk 4]
```diff
@@ -578,11 +580,13 @@ export default class Dropdown extends Component {
...
   selectItemOnEnter = (e) => {
-    debug('selectItemOnEnter()')
-    debug(keyboardKey.getName(e))
+    debug('selectItemOnEnter()', keyboardKey.getName(e))
+    const { search } = this.props
+
     if (keyboardKey.getCode(e) !== keyboardKey.Enter) return
     e.preventDefault()
 
+    if (search && _.isEmpty(this.getMenuOptions())) return
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:589]: Disables `Enter` key when nothing find.

File: src/modules/Dropdown/Dropdown.js [hunk 5]
```diff
@@ -622,29 +626,34 @@ export default class Dropdown extends Component {
...
     debug('handleMouseDown()')
-    const { onMouseDown } = this.props
-    if (onMouseDown) onMouseDown(e, this.props)
+
     this.isMouseDown = true
+    _.invoke(this.props, 'onMouseDown', e, this.props)
     // Do not access document when server side rendering
-    if (!isBrowser) return
-    document.addEventListener('mouseup', this.handleDocumentMouseUp)
+    if (isBrowser) document.addEventListener('mouseup', this.handleDocumentMouseUp)
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:633]: Minor refactor.

File: src/modules/Dropdown/Dropdown.js [hunk 6]
```diff
@@ -622,29 +626,34 @@ export default class Dropdown extends Component {
...
   }
 
   handleDocumentMouseUp = () => {
     debug('handleDocumentMouseUp()')
+
     this.isMouseDown = false
     // Do not access document when server side rendering
-    if (!isBrowser) return
-    document.removeEventListener('mouseup', this.handleDocumentMouseUp)
+    if (isBrowser) document.removeEventListener('mouseup', this.handleDocumentMouseUp)
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:641]: Minor refactor.

File: src/modules/Dropdown/Dropdown.js [hunk 7]
```diff
@@ -622,29 +626,34 @@ export default class Dropdown extends Component {
...
+    const { open, searchQuery } = this.state
+
+    _.invoke(this.props, 'onClick', e, this.props)
     // prevent closeOnDocumentClick()
     e.stopPropagation()
-    this.toggle(e)
+
+    if (!search) return this.toggle(e)
+    if (open) return
+    if (searchQuery.length >= minCharacters || minCharacters === 1) this.open(e)
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:656]: Update behaviour with click on search input.

File: test/specs/modules/Dropdown/Dropdown-test.js [hunk 8]
```diff
@@ -1618,19 +1712,6 @@ describe('Dropdown', () => {
...
-
-      dropdownMenuIsClosed()
-
-      // simulate search with query's length is less than minCharacters
-      wrapper
-        .find('input.search')
-        .simulate('change', { target: { value: faker.hacker.noun().substring(0, 1) } })
-
-      dropdownMenuIsClosed()
-    })
```
Comment (layershifter) [inline test/specs/modules/Dropdown/Dropdown-test.js:1632]: Test was moved to another section.

**Review by layershifter — COMMENTED** [review]
@levithomason We're ready for review there, will be awesome to hear your thoughts.

**Comment by timworx** [comment]
Would be great to see this merged. 

It fixes a considerable bug in `handleSearchChange`. Within that function `setState` happens in an if block that keeps it from being called when you're deleting the last character in the `searchQuery`.

File: src/modules/Dropdown/Dropdown.js [hunk 9]
```diff
@@ -1013,22 +1023,7 @@ export default class Dropdown extends Component {
...
-    const options = this.getMenuOptions()
-
-    if (search && _.isEmpty(options)) {
-      e.preventDefault()
-      return
-    }
-
-    this.close(e)
-  }
+  toggle = e => this.state.open ? this.close(e) : this.open(e)
```
Comment (layershifter) [inline src/modules/Dropdown/Dropdown.js:1026]: Method was updated because its logic was moved to `handleClick`

**Comment by layershifter** [comment]
@timworx thanks for the feedback, I missed this problem, hovewer it's fixed.

Just a note, @levithomason I think you should ship this in `0.68.6`, because it fixes critical bugs, without this PR the `Dropdown` with a `search` prop is completely unusable.

**Comment by layershifter** [comment]
@levithomason please take an attention to this PR.

**Comment by levithomason** [comment]
The next shipment (here in a few minutes) will be breaking because of #1730 which has already been merged.  We won't update previous versions until we hit a stable 1.0 release.  Bugs and features will be shipped as fast as possible without going backwards.

I'll demo this and see about shipping it now in `0.69.0`.

**Comment by levithomason** [comment]
Released in `semantic-ui-react@0.69.0`

**Pull request merged**